### DISCUSSION TITLE
Assigning ranks to members

### DISCUSSION THREAD

**Discussion by enrico-ghidoni** [CntoDev/level-up#84] [Ideas]
Make all Ranks available (Recruit, Reservist, Grunt, Specialist, Corporal and Staff Sergeant) for promoting/demoting members.

Different ranks can be assigned by different people, thus the permission system should be updated accordingly.

I am not currently aware whether the community already has guidelines for which ranks/branch members can assign which ranks to other members (pinging @JamesTheClarke here to make sure). But ideally it should be something like this:
 - only Interviewers and SSgts can promote a Recruit to the Reservist rank
 - only Member Managers and SSgts can promote/demote Reservists, Grunts and Specialists to either three of those ranks (although a valid alternative would be to make only Branch managers be able to promote members to the Specialist rank, but that would require the Branch system to be in place so it can wait)
 - only SSgts can promote members to Corporal and Staff Sergeant

**Comment by milivojm** [comment]
I'd rephrase this to:

- SSgts can promote Member to Corporal
- Branch Managers can promote Member to Specialists
- Members voting for SSgt promotion

Grunts, Reservists, Recruits will be managed automatically.

**Comment by enrico-ghidoni** [comment]
Moving discussion about automation to #68

**Comment by milivojm** [comment]
Maybe this is something we cannot do much about, it's a user command. It is linked with #60

**Comment by enrico-ghidoni** [comment]
For automation probably not. But the idea of the ranks system is heavily tied to permissions within the Roster, although the latter can be skipped for MVP in my opinion

**Comment by milivojm** [comment]
Yes, it's best to create a consumer in permissions service that listens to events like MemberPromoted and give appropriate permissions. 

Right now we have no domain concept of Branch. We have Ranks which are tightly controlled by fixed values. 

Work in progress.

**Comment by enrico-ghidoni** [comment]
Branches shouldn't be in the MVP I think. Although we'll have both Interviewers and Member Managers using new Roster so that could be an issue.

But then again, let's not fiddle with permissions right now and only handle the events for Rank promotion/demotion.

**Comment by milivojm** [comment]
Interviewing promotes from RCT to RES.
SSGT promote SPC, CPL, SSGT.

System could automatically promote/demote RES/GNT based on attendance data.

**Comment by milivojm** [comment]
@JamesTheClarke suggested CPL rank could be used to promote members into SPC for their own branch. That implies several things that need to be done before ranks:
- Identity context needs to handle `MemberPromoted` events by giving/taking away [claims](https://github.com/CntoDev/rooster/blob/develop/src/Roster.Web/Security/Policy.cs). See next bulletin for details.
- We need to create some new claims: 
  - `PromoteStaffSergeant` - adds ability to promote someone to SSGT. This claim would be given to Member who has been promoted to SSGT.
  - `PromoteCorporal` - adds ability to promote someone to CPL. This claim would be given to Member who has been promoted to SSGT.
  - `PromoteSpecialist` - adds ability to promote someone to SPC. This claim would be given to Member who has been promoted to SSGT or CPL.
- Branch concept needs to be introduced because CPL can only promote people to their own branch.
- Event management context will need to publish a new event which would represent attendance ratio, for example `MemberAttendedEvents`. That would trigger promoting/demoting GNT. It would contain properties:
  - `Nickname` - member nickname
  - `EventsAttended` - number of events attended in last tracking window (think it's 2 months as we speak)
  - `GruntQuotaSatisfied` - true if enough events have been attended by given member

**Comment by enrico-ghidoni** [comment]
It depends on what we want to have for MVP I think. We have a couple of options:
 1. We stick to `django-roster` user system, where users are only SSgts, Interviewers and Member Manager thus we give all registered users the corresponding *claims* for promoting/demoting a member. We can avoid implementing the concept of *branch* which will be addressed in future releases, it would just require manual setup of the *claims* you mentioned and the ranks system.
 2. We implement the **branch** system before MVP release, keeping in mind that still registered users are only going to be SSgts, Interviewers and Member Managers. This obviously requires more time to develop but could give some feedback about the design of the *claim* system that you just described (which sounds valid to me).

Concerning your last point about managing *community events* I don't agree with including the `GruntQuotaSatisfied` field in the event itself as it should be in a separate module. Although this isn't the right place to discuss the issue, I'll post something more detailed in #85.

**Reply by milivojm** [reply]
> It depends on what we want to have for MVP I think. We have a couple of options:
> 
> 1. We stick to `django-roster` user system, where users are only SSgts, Interviewers and Member Manager thus we give all registered users the corresponding _claims_ for promoting/demoting a member. We can avoid implementing the concept of _branch_ which will be addressed in future releases, it would just require manual setup of the _claims_ you mentioned and the ranks system.

Doable. Would require manual intervention occasionally. Also, CPLs would need to be fair and not promote people from other branches.

**Reply by enrico-ghidoni** [reply]
If we only have SPC+ from only two branches (Interview and Member Management) I don't think the conflict you mention will ever happen. Without any concept of *branch membership* a CPL cannot promote a member of the other branch since they are SPC already.

Regarding the manual intervention, this solution is only temporary and any manual change to the database will require less time than implementing the whole system anyway so I'm good with the solution.

**Comment by milivojm** [comment]
Sounds good then, this is ready for breakdown.

**Comment by milivojm** [comment]
Opened and implemented issue #100
